[Merged by Bors] - refactor(SimpleGraph): change bridges not to require the edge to be present

MJ141592 opened the pull request:
Remove the requirement for edges to exist in definition of isBridge, and then change the related lemmas to adjust for this,  by renaming, fixing the statements and fixing the proofs, including in Acyclic.

Closes #31690.

---

- [x] depends on: #36802
- [x] depends on: #36804

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -791,23 +784,22 @@ theorem adj_and_reachable_delete_edges_iff_exists_cycle {v w : V} :
+    rw [← adj_and_reachable_delete_edges_iff_exists_cycle]
+    simp only [hvw, true_and]
+  induction e
+  apply h he
```
```suggestion
  obtain ⟨v, w⟩ := e
  rw [isBridge_iff]
  contrapose!
  rw [← adj_and_reachable_delete_edges_iff_exists_cycle]
  simp_all
```

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -717,11 +712,17 @@ section BridgeEdges
-    G.IsBridge s(u, v) ↔ G.Adj u v ∧ ¬(G \ fromEdgeSet {s(u, v)}).Reachable u v := Iff.rfl
+    G.IsBridge s(u, v) ↔ ¬(G \ fromEdgeSet {s(u, v)}).Reachable u v := Iff.rfl
+
+theorem not_edge_isBridge (e : Sym2 V) (hne : ¬(e ∈ G.edgeSet)) : G.IsBridge e := by
```
```suggestion
theorem isBridge_of_not_mem_edgeSet {e : Sym2 V} (he : e \nin G.edgeSet) : G.IsBridge e := by
```
but I don't think this result is true. This will depend on whether there is another path between the endpoints of `z`

SnirBroshi commented:
Hello 👋
Please merge master (click "Update branch" below), there are 2 new lemmas about bridges from #32870 that would need adjusting for the new definition

SnirBroshi commented:
It's unfortunate, but I think there's no rush on this, unless I've missed some PR that needs this.
I was thinking of offering help, but if you're up to this and it helps you learn Lean then leaving it to you sounds good.

YaelDillies committed:
refactor(SimpleGraph): change bridges not to require the edge to be present
Author: MJ141592 <95647331+MJ141592@users.noreply.github.com>
Date:   Tue Mar 17 15:59:50 2026 +0100
Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

YaelDillies committed:
deprecated aliases

Rida-Hamadani commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -743,17 +738,16 @@ lemma Preconnected.exists_adj_of_nontrivial [Nontrivial V] {G : SimpleGraph V} (
 def IsBridge (G : SimpleGraph V) (e : Sym2 V) : Prop :=
-  e ∈ G.edgeSet ∧
-    Sym2.lift ⟨fun v w => ¬(G.deleteEdges {e}).Reachable v w, by simp [reachable_comm]⟩ e
+  Sym2.lift ⟨fun v w => ¬(G.deleteEdges {e}).Reachable v w, by simp [reachable_comm]⟩ e
```
```suggestion
  Sym2.lift ⟨fun v w ↦ ¬ (G.deleteEdges {e}).Reachable v w, by simp [reachable_comm]⟩ e
```

Rida-Hamadani commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -775,10 +769,12 @@ theorem reachable_deleteEdges_iff_exists_walk {v w v' w' : V} :
-  rw [reachable_deleteEdges_iff_exists_walk, not_exists_not]
+theorem isBridge_iff_forall_walk_mem_edges {v w : V} :
+    G.IsBridge s(v, w) ↔ ∀ p : G.Walk v w, s(v, w) ∈ p.edges := by
+  rw [isBridge_iff,reachable_deleteEdges_iff_exists_walk, not_exists_not]
```
```suggestion
  rw [isBridge_iff, reachable_deleteEdges_iff_exists_walk, not_exists_not]
```

Rida-Hamadani commented on Mathlib/Combinatorics/SimpleGraph/Acyclic.lean:
```diff
@@ -164,23 +164,17 @@ theorem IsTree.coe_subgraphOfAdj {u v : V} (h : G.Adj u v) : G.subgraphOfAdj h |
-    G.IsAcyclic ↔ ∀ ⦃e⦄, e ∈ (G.edgeSet) → G.IsBridge e := by
-  simp [isAcyclic_iff_forall_adj_isBridge, Sym2.forall]
+theorem isAcyclic_iff_forall_isBridge : G.IsAcyclic ↔ ∀ ⦃e⦄, e ∈ G.edgeSet → G.IsBridge e where
+  mp hG e he := by
```
I think `refine` here would be better than `where` notation.

YaelDillies replied:
Why? Do you generally dislike `where` notation for iffs? At any rate, this is a style concern, so not worth bringing up in review unless the style guide has an opinion

Rida-Hamadani replied:
It costs one less line and it is consistent with the rest of the file (also I have seen maintainers ask to not use `where` for mere iff statements in other reviews).

But I am planning to review again addressing non-style issues (if I find any), this was just a shallow first review.

SnirBroshi replied:
> (also I have seen maintainers ask to not use `where` for mere iff statements in other reviews)

FWIW [this](https://github.com/leanprover-community/mathlib4/pull/22657#discussion_r2029970925) is one example, and Jireh's review in #34395 is an opposite example.

YaelDillies committed:
apace

Co-authored-by: Rida Hamadani <mridahamadani@gmail.com>

YaelDillies committed:
unicode

Co-authored-by: Rida Hamadani <mridahamadani@gmail.com>

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into MJ_IsBridge_31690

Rida-Hamadani commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -835,23 +831,29 @@ theorem adj_and_reachable_delete_edges_iff_exists_cycle {v w : V} :
-  rw [isBridge_iff, and_congr_right_iff]
-  intro h
+theorem isBridge_iff_forall_cycle_notMem {e : Sym2 V} (he : e ∈ G.edgeSet) :
+    G.IsBridge e ↔ ∀ ⦃u : V⦄ (p : G.Walk u u), p.IsCycle → e ∉ p.edges := by
```
The proof could be shortened to
```lean4
  obtain ⟨_, _⟩ := e
  contrapose
  simp_all [isBridge_iff, ← adj_and_reachable_delete_edges_iff_exists_cycle]
```

YaelDillies replied:
Do you know you can write multiline commands on github?

Rida-Hamadani replied:
Yes of course, I decided not to do that since the lines to be changed here had non-diff lines in the middle.

Rida-Hamadani commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -862,29 +864,32 @@ lemma Connected.connected_delete_edge_of_not_isBridge (hG : G.Connected) {x y :
-    (G ⊔ edge u v).IsBridge s(u, v) := by
-  refine isBridge_iff.mpr ⟨.inr ⟨Set.mem_singleton _, mt (· ▸ .rfl) h⟩, ?_⟩
-  exact fun h' ↦ h <| .mono (sdiff_le_iff'.mpr <| refl _) h'
+    (G ⊔ edge u v).IsBridge s(u, v) := by simpa using .of_not_reachable h
```
Since `simpa` is a finishing tactic, it cannot always be inlined
```suggestion
    (G ⊔ edge u v).IsBridge s(u, v) := isBridge_sup_edge.mpr (of_not_reachable h)
```

Rida-Hamadani commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/EdgeConnectivity.lean:
```diff
@@ -123,18 +123,20 @@ lemma isEdgeConnected_add_one (hk : k ≠ 0) :
   simp [IsEdgeConnected, isEdgeReachable_add_one hk, forall_comm (α := Sym2 _)]
 
 /-- An edge is a bridge iff its endpoints are adjacent and not 2-edge-reachable. -/
```
```suggestion
/-- An edge is a bridge iff its endpoints are not 2-edge-reachable. -/
```

Rida-Hamadani commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/EdgeConnectivity.lean:
```diff
@@ -123,18 +123,20 @@ lemma isEdgeConnected_add_one (hk : k ≠ 0) :
+  by_cases! hs₁ : s.encard ≠ (1 : ℕ)
+  · apply G.isEdgeReachable_one.mpr huv.reachable
+    exact lt_of_le_of_ne (ENat.lt_coe_add_one_iff.mp hs₂) hs₁
+  obtain ⟨x, rfl⟩ := Set.encard_eq_one (s := s).mp hs₁
```
```suggestion
  obtain ⟨x, rfl⟩ := s.encard_eq_one.mp hs₁
```

YaelDillies committed:
golf

SnirBroshi commented on Mathlib/Combinatorics/SimpleGraph/Acyclic.lean:
```diff
@@ -170,23 +170,17 @@ theorem IsTree.coe_subgraphOfAdj {u v : V} (h : G.Adj u v) : G.subgraphOfAdj h |
   simp_all
   grind [IsCycle.snd_ne_penultimate]
 
-theorem isAcyclic_iff_forall_adj_isBridge :
```
Missing deprecation

SnirBroshi commented on Mathlib/Combinatorics/SimpleGraph/Acyclic.lean:
```diff
@@ -170,23 +170,17 @@ theorem IsTree.coe_subgraphOfAdj {u v : V} (h : G.Adj u v) : G.subgraphOfAdj h |
+  mp hG e he := by
+    rw [isBridge_iff_forall_cycle_notMem he]
+    intro hu hp hc
+    cases hG hp hc
```
```suggestion
  mp hG e he := isBridge_iff_forall_cycle_notMem he |>.mpr fun v p hc ↦ hG p hc |>.elim
```

SnirBroshi commented on Mathlib/Combinatorics/SimpleGraph/Acyclic.lean:
```diff
@@ -170,23 +170,17 @@ theorem IsTree.coe_subgraphOfAdj {u v : V} (h : G.Adj u v) : G.subgraphOfAdj h |
+    intro hu hp hc
+    cases hG hp hc
+  mpr hG v c hc := by
+    obtain ⟨e, he⟩ := c.edges.exists_mem_of_ne_nil (by simp [hc.ne_nil, nil_iff_eq_nil])
```
```suggestion
    obtain ⟨e, he⟩ := c.edges.exists_mem_of_ne_nil <| by simp [hc.not_nil]
```

SnirBroshi commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/EdgeConnectivity.lean:
```diff
@@ -122,19 +122,21 @@ lemma isEdgeConnected_add_one (hk : k ≠ 0) :
-    exact deleteEdges_adj.mpr ⟨hadj, hx⟩ |>.reachable
+/-- An edge is a bridge iff its endpoints are not 2-edge-reachable. -/
+lemma isBridge_iff_not_isEdgeConnected_two {u v : V} (huv : G.Adj u v) :
+    G.IsBridge s(u, v) ↔ ¬G.IsEdgeReachable 2 u v := by
```
I think the left-to-right implication is still true without `G.Adj u v`, can you separate that into another lemma?
I also think we should have `G.IsBridge s(u, v) → G.Reachable u v → G.Adj u v` which should be useful there.

YaelDillies replied:
Not sure how that second lemma would be useful here

SnirBroshi replied:
My thinking was that you could `by_cases G.Adj u v` and then that lemma would solve the negative branch (when combined with `G.IsEdgeReachable 2 u v → G.Reachable u v`), but I didn't notice the existing left-to-right proof didn't use the `G.Adj` hypothesis.

I think that lemma is useful anyways, can you add it?

YaelDillies replied:
Sure, but I have already spent a lot of time splitting features and chores off this refactor, so I do not want to add anything that is unnecessary. Happy for you to add it in a later PR of course.

SnirBroshi commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/EdgeConnectivity.lean:
```diff
@@ -122,19 +122,21 @@ lemma isEdgeConnected_add_one (hk : k ≠ 0) :
-    · exact hx ▸ hr
-    exact deleteEdges_adj.mpr ⟨hadj, hx⟩ |>.reachable
+/-- An edge is a bridge iff its endpoints are not 2-edge-reachable. -/
+lemma isBridge_iff_not_isEdgeConnected_two {u v : V} (huv : G.Adj u v) :
```
```suggestion
lemma Adj.isBridge_iff_not_isEdgeConnected_two {u v : V} (huv : G.Adj u v) :
```

YaelDillies replied:
I am not convinced here. This lemma isn't primarily about adjacency

SnirBroshi commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -29,11 +29,6 @@ public import Mathlib.Combinatorics.SimpleGraph.Operations
 * `SimpleGraph.isBridge_iff_mem_and_forall_cycle_notMem` characterizes bridge edges in terms of
   there being no cycle containing them.
```
```suggestion
* `SimpleGraph. isBridge_iff_forall_cycle_notMem` characterizes bridge edges in terms of there being
  no cycle containing them.
```

SnirBroshi commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -835,23 +831,27 @@ theorem adj_and_reachable_delete_edges_iff_exists_cycle {v w : V} :
+  fun hep ↦ (isBridge_iff_forall_cycle_notMem <| p.edges_subset_edgeSet hep).mp he _ hp hep
+
+@[deprecated (since := "2026-04-02")]
+alias isBridge_iff_mem_and_forall_cycle_notMem := IsBridge.notMem_edges_of_isCycle
```
```suggestion
alias isBridge_iff_mem_and_forall_cycle_notMem := isBridge_iff_forall_cycle_notMem
```

SnirBroshi commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -862,29 +862,32 @@ lemma Connected.connected_delete_edge_of_not_isBridge (hG : G.Connected) {x y :
 
 /-- If `e` is an edge in `G` and is a bridge in a larger graph `G'`, then it's a bridge in `G`. -/
 theorem IsBridge.anti_of_mem_edgeSet {G' : SimpleGraph V} {e : Sym2 V} (hle : G ≤ G')
     (h : e ∈ G.edgeSet) (h' : G'.IsBridge e) : G.IsBridge e :=
```
`h` is now redundant, and this should be called `IsBridge.anti`

YaelDillies committed:
Snir's suggestions

YaelDillies committed:
fix

SnirBroshi commented on Mathlib/Combinatorics/SimpleGraph/Acyclic.lean:
```diff
@@ -170,23 +170,17 @@ theorem IsTree.coe_subgraphOfAdj {u v : V} (h : G.Adj u v) : G.subgraphOfAdj h |
+    exact (hG <| c.edges_subset_edgeSet he).notMem_edges_of_isCycle hc he
+
+@[deprecated (since := "2026-05-16")]
+alias isAcyclic_iff_forall_adj_edge_isBridge := isAcyclic_iff_forall_isBridge
```
typo
```suggestion
alias isAcyclic_iff_forall_adj_isBridge := isAcyclic_iff_forall_isBridge
```

SnirBroshi commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -26,13 +26,8 @@ public import Mathlib.Combinatorics.SimpleGraph.Operations
-
-`IsBridge` is unpractical: we shouldn't require the edge to be present.
-See https://github.com/leanprover-community/mathlib4/issues/31690.
+* `SimpleGraph. isBridge_iff_forall_cycle_notMem` characterizes bridges as the edges not
```
```suggestion
* `SimpleGraph.isBridge_iff_forall_cycle_notMem` characterizes bridges as the edges not
```
(typo originates from my suggestion, sorry)

SnirBroshi commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -862,29 +862,30 @@ lemma Connected.connected_delete_edge_of_not_isBridge (hG : G.Connected) {x y :
+@[deprecated (since := "2026-05-16")] alias IsBridge.anti_of_mem_edgeSet := IsBridge.anti
+
+@[simp]
+lemma isBridge_sup_edge {u v : V} : (G ⊔ edge u v).IsBridge s(u, v) ↔ G.IsBridge s(u, v) := by
```
Can you also add `(G.deleteEdges e).IsBridge e ↔ G.IsBridge e`?

YaelDillies committed:
typos

YaelDillies committed:
less variables

SnirBroshi reviewed (approved):
Thanks!

YaelDillies committed:
Squashed commit of the following:

commit bc8c3b3d8e5a2f1165245ed314fde0b6c71eb71a
Author: Yaël Dillies <yael.dillies@gmail.com>
Date:   Sat May 16 18:33:50 2026 +0200

    less variables

commit fbbef302be4a659c90323130e3806c930d5cc7c5
Author: Yaël Dillies <yael.dillies@gmail.com>
Date:   Sat May 16 18:26:56 2026 +0200

    typos

commit af551f6d9e8a011feef6323da43611b822f8ad2b
Author: Yaël Dillies <yael.dillies@gmail.com>
Date:   Sat May 16 13:58:25 2026 +0200

    fix

commit eb24e1be1246653a9270a873eebc3767ef057304
Author: Yaël Dillies <yael.dillies@gmail.com>
Date:   Sat May 16 13:57:46 2026 +0200

    Snir's suggestions

commit fe5514f3017d11fc113ae41c706f1ce90ac2bb48
Author: Yaël Dillies <yael.dillies@gmail.com>
Date:   Fri May 15 10:49:36 2026 +0200

    golf

commit d580d76e11a6ff46f3fc57cdfb51234666f6fc20
Merge: 4f3209903be 1c1dadbc285
Author: Yaël Dillies <yael.dillies@gmail.com>
Date:   Wed May 13 08:40:44 2026 +0200

    Merge remote-tracking branch 'origin/master' into MJ_IsBridge_31690

commit 4f3209903befd2bc47f972933b16cc3ade46d5c5
Author: Yaël Dillies <yael.dillies@gmail.com>
Date:   Wed May 13 05:57:13 2026 +0200

    unicode

    Co-authored-by: Rida Hamadani <mridahamadani@gmail.com>

commit 966b19966d954db8470a976e549a7329f7df458e
Author: Yaël Dillies <yael.dillies@gmail.com>
Date:   Wed May 13 05:56:57 2026 +0200

    apace

    Co-authored-by: Rida Hamadani <mridahamadani@gmail.com>

commit 7c346745cafa048986786d6d1656343a0a24f859
Author: Yaël Dillies <yael.dillies@gmail.com>
Date:   Thu Apr 2 09:00:09 2026 +0000

    deprecated aliases

commit 07a3c843b63a70709db9c541a22b006404612871
Author: Yaël Dillies <yael.dillies@gmail.com>
Date:   Thu Apr 2 08:58:57 2026 +0000

    refactor(SimpleGraph): change bridges not to require the edge to be present
    Author: MJ141592 <95647331+MJ141592@users.noreply.github.com>
    Date:   Tue Mar 17 15:59:50 2026 +0100
    Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

b-mehta commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -862,29 +862,33 @@ lemma Connected.connected_delete_edge_of_not_isBridge (hG : G.Connected) {x y :
+  induction e with | h u v; simp [isBridge_iff]
 
-/-- Connecting two unreachable vertices by an edge creates a bridge. -/
+@[deprecated "Inline the proof" (since := "2026-04-02")]
```
What does this deprecation message mean? It seems incredibly unhelpful and unfriendly to downstream contributors.

YaelDillies replied:
This means "Look at the proof, and repeat what it does". Any other deprecation message I could think of is less precise than that, but I've switched anyway

b-mehta replied:
Your new one is significantly more precise, thanks!

b-mehta commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -862,29 +862,33 @@ lemma Connected.connected_delete_edge_of_not_isBridge (hG : G.Connected) {x y :
 @[deprecated (since := "2026-03-18")]
 alias IsBridge.sup_fromEdgeSet_of_not_reachable := IsBridge.sup_edge_of_not_reachable
 
 /-- Connecting two unreachable vertices by an edge preserves existing bridges. -/
```
```suggestion
/-- Connecting two unreachable vertices by an edge preserves existing bridges,
provided the bridge is already an edge. -/
```

b-mehta commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/EdgeConnectivity.lean:
```diff
@@ -122,19 +122,23 @@ lemma isEdgeConnected_add_one (hk : k ≠ 0) :
+lemma IsBridge.not_isEdgeReachable_two (huv : G.IsBridge s(u, v)) : ¬ G.IsEdgeReachable 2 u v :=
+  fun hc ↦ huv <| hc <| Set.encard_singleton _ |>.trans_lt Nat.one_lt_ofNat
+
+/-- An edge is a bridge iff its endpoints are not 2-edge-reachable. -/
```
```suggestion
/-- An edge is a bridge iff its endpoints are not 2-edge-reachable. 
The forward direction of this is true without assuming `u` and `v` are adjacent, see `IsBridge.not_isEdgeReachable_two`. -/
```

b-mehta commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/EdgeConnectivity.lean:
```diff
@@ -122,19 +122,23 @@ lemma isEdgeConnected_add_one (hk : k ≠ 0) :
+  · apply G.isEdgeReachable_one.mpr huv.reachable
+    exact lt_of_le_of_ne (ENat.lt_coe_add_one_iff.mp hs₂) hs₁
+  obtain ⟨x, rfl⟩ := s.encard_eq_one.mp hs₁
+  by_cases hx : s(u, v) = x
```
Using `obtain rfl | hx := eq_or_ne s(u, v) x` would be nice

b-mehta commented on Mathlib/Combinatorics/SimpleGraph/Acyclic.lean:
```diff
@@ -372,7 +366,7 @@ lemma IsTree.card_edgeFinset [Fintype V] [Fintype G.edgeSet] (hG : G.IsTree) :
 /-- A minimally connected graph is a tree. -/
 lemma isTree_of_minimal_connected (h : Minimal Connected G) : IsTree G := by
-  rw [isTree_iff, and_iff_right h.prop, isAcyclic_iff_forall_adj_isBridge]
+  rw [isTree_iff, and_iff_right h.prop, isAcyclic_iff_forall_isBridge, Sym2.forall]
```
This strongly suggests that removing the `isAcyclic` lemma which uses `G.Adj` instead of `G.edgeSet` on the rhs is an api regression.

b-mehta reviewed (commented):
Don't forget to update deprecations.

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into MJ_IsBridge_31690

YaelDillies committed:
Bhavik's suggestions

b-mehta commented on Mathlib/Combinatorics/SimpleGraph/Acyclic.lean:
```diff
@@ -512,7 +505,7 @@ lemma isTree_iff_connected_and_card [Finite V] :
   refine ⟨fun h ↦ ⟨h.connected, by simpa [edgeFinset] using h.card_edgeFinset⟩,
     fun ⟨h₁, h₂⟩ ↦ ⟨h₁, ?_⟩⟩
-  simp_rw [isAcyclic_iff_forall_adj_isBridge]
+  simp_rw [isAcyclic_iff_forall_isBridge, Sym2.forall]
```
```suggestion
  simp_rw [isAcyclic_iff_forall_adj_isBridge]
```
untested

b-mehta reviewed (commented):
Thanks!

bors d+

YaelDillies committed:
revert

Co-authored-by: Bhavik Mehta <bhavikmehta8@gmail.com>

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into MJ_IsBridge_31690

YaelDillies committed:
fix
